[RFC] Create Threat Fieldset - Stage 2 Proposal (#1293)
elastic/ecs: rfcs/text/0008-threat-intel.md

threat.indicator.first_seen | date | 2020-12-01 | The date and time when intelligence source first reported sighting this indicator
threat.indicator.last_seen | date | 2020-12-02| The date and time when intelligence source last reported sighting this indicator.
threat.indicator.sightings | long | 20 | Number of times this indicator was observed conducting threat activity
threat.indicator.type | keyword | ipv4-addr | Type of indicator as represented by Cyber Observable in STIX 2.0
threat.indicator.description | keyword | 201.10.10.90 was seen delivering Angler EK | Describes the type of action conducted by the threat

> threat.indicator.dataset | keyword | threatintel.{abusemalware,abuseurl,misp,otx,limo} | Identifies the name of specific dataset from the intelligence source.

Reviewer: @peasead is this one (and some others, I think) redundant with e.g. `event.dataset`, or does this serve a separate purpose?

Author: Removed this. Thanks.

threat.indicator.ip | ip | 1.2.3.4 | Identifies a threat indicator as an IP address (irrespective of direction).
threat.indicator.domain | keyword | evil.com | Identifies a threat indicator as a domain (irrespective of direction).
threat.indicator.port | long | 443 | Identifies a threat indicator as a port number (irrespective of direction).
threat.indicator.email.address | keyword | phish@evil.com | Identifies a threat indicator as an email address (irrespective of direction).
threat.marking.tlp | keyword | RED | Data markings represent restrictions, permissions, and other guidance for how data can be used and shared. Examples could be TLP (WHITE, GREEN, AMBER, RED).
threat.indicator.scanner_stats | long | 4 | Count of Anti virus/EDR that successfully detected malicious file or URL. Sources like VirusTotal, Reversing Labs often provide these statistics.

### Proposed New Values for Event Fieldset

Field | New Value | Description
--- | --- | ---

> event.kind | enrichment | Propose adding this value to capture the type of information this event contains and how it should be used. Threat intelligence will be used to enrich source events and signals. Enrichment could also apply to other types of contextual data sources (not just threat intelligence) such as directory services, IPAM data, asset lists.

Reviewer 2: Are these new proposed values for `event.*` fields still in the scope of this proposal?

Reviewer: @ebeahan while we don't currently have a definite use case for e.g. `event.kind: enrichment` (since indicator match rules generate signals, not enrichments), I think having this new kind makes sense if one wanted an audit trail for enrichment, and similarly I don't think the other fields are still relevant as well 👍

threat.indicator.first_seen | date | 2020-12-01 | The date and time when intelligence source first reported sighting this indicator
threat.indicator.last_seen | date | 2020-12-02| The date and time when intelligence source last reported sighting this indicator.
threat.indicator.sightings | long | 20 | Number of times this indicator was observed conducting threat activity
threat.indicator.type | keyword | ipv4-addr | Type of indicator as represented by Cyber Observable in STIX 2.0
threat.indicator.description | keyword | 201.10.10.90 was seen delivering Angler EK | Describes the type of action conducted by the threat

> threat.indicator.dataset | keyword | threatintel.{abusemalware,abuseurl,misp,otx,limo} | Identifies the name of specific dataset from the intelligence source.

Author: This should just be `event.module`. As an example, for the Filebeat TI module, it would be `event.module:threatintel`.

Typo and I'll fix it.

> threat.indicator.module | keyword | threatintel | Identifies the name of specific module where the data is coming from.

Author: Yes, this should just be `event.dataset:threatintel.{abusemalware,abuseurl,misp,otx,limo}`.

I'll get that fixed.

threat.indicator.provider | keyword | Abuse.ch | Identifies the name of intelligence provider.

> threat.indicator.reference | keyword | https://urlhaus.abuse.ch/url/1292596 | Provides a reference for the indicator for additional information.

Author: If we're moving all of the `event.` data up, we should move `event.reference` up as well.

Example: `event.reference:https://urlhaus.abuse.ch/url/1321083/`

While there are two primary uses for these fields, this RFC deals primarily with the first: ingestion/storage of threat intelligence.

**Storing threat intelligence as an event document in threat index(s).**

Threat intelligence data will be collected from multiple sources stored in threat indices. The ECS fields proposed here will be used to structure the documents collected from various sources.

**Examples**

Network Example
```json5
{
    // Metadata about the indicator event
    "@timestamp": "2019-08-10T11:09:23.000Z",
    "event": {
        "kind": "enrichment",
        "category": "threat",
        "type": "indicator",
        "reference": "https://urlhaus.abuse.ch",
        "severity": 7,
        "risk_score": 10,
        "original": "2020-10-29 19:16:38"
    },

    // Metadata about the indicator data
    "threat.indicator": {

>         "first_seen": "2020-10-01",

Reviewer: @peasead it's probably outside the scope of this RFC, but I wanted to mention that the need for an `updated_at` field has been requested in the context of indicator indexes, in order to be able to query the most relevant indicators by sorting on said field. I think that this goes against our current idea of indicators as timeseries data, but the underlying problem can be expressed as three events, in chronological order:

1. Domain indicator A is active
2. Domain indicator B is active
3. Domain indicator A is inactive

If the analyst is only concerned with active domains, then a naive query may include both active and inactive versions of indicator A if scoping simply on `@timestamp` or `event.ingested`.

Author: I think we should add it now.

I'll do that.

Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
-->

The additions described above will be used to enable cyber threat intelligence capabilities in Elastic Security solution. A new rule type Indicator match was introduced in 7.10 and the proposed ECS updates will enable a new category of detection alerts that match incoming log and event data against threat intelligence sources. Additionally in the future we will also develop enrichment flows that add context from threat intelligence to alerts and events to assist analysts in their investigative workflows.

While there are two primary uses for these fields, this RFC deals primarily with the first: ingestion/storage of threat intelligence.

**Storing threat intelligence as an event document in threat index(s).**

    Threat intelligence data will be collected from multiple sources stored in threat indices. The ECS fields proposed here will be used to structure the documents collected from various sources.

> 

Reviewer: @peasead the example below declares `indicator.*` instead of `threat.indicator.*`, and as mentioned by @ebeahan is also missing many of the fields in the above table.

Author: Yep. I'm working that.

Author: yessir. I got that buttoned up.

        "first_seen": "2020-10-01",
        "last_seen": "2020-11-01",
        "sightings": "10",
        "type": [
            "ipv4-addr",
            "port",
            "domain-name",
            "email-addr"
        ],
        "description": "Email address, domain, port, and IP address observed using an Angler EK campaign.",
        "dataset": "threatintel.abuseurl",
        "module": "threatintel",
        "provider": "Abuse.ch",
        "reference": "https://urlhaus.abuse.ch/url/1292596/",
        "confidence": "High",
        "ip": "1.2.3.4",
        "domain": "malicious.evil",
        "port": 443,
        "email.address": "phish@malicious.evil",
        "marking.tlp": "WHITE",
        "scanner_stats": 4
    },

    // Any indicators should also be copied to relevant related.* field
    "related": {
        "hash": [
            "1eee2bf3f56d8abed72da2bc523e7431",
            "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4"
        ],
        "hosts": [
            "nefarious.evil"
        ],
        "ip": [
            "1.2.3.4"
        ]
    },

    // Tags for context
    "tags": [
        "threatintel",
        "forwarded"
    ],
    ]
}

> ```

Author: Doh. Sorry about that!

        ],
        "hosts": [
            "nefarious.evil"
        ],
        "ip": [
            "1.2.3.4"
        ]
    },

    // Tags for context
    "tags": [
        "threatintel",
        "forwarded"
    ]
}
```

File Example
```json5
{
    // Metadata about the indicator event
    "@timestamp": "2019-08-10T11:09:23.000Z",
    "event": {
        "kind": "enrichment",
        "category": "threat",
        "type": "indicator",
        "reference": "https://bazaar.abuse.ch",
        "severity": 7,
        "risk_score": 10,
        "original": "2020-10-29 19:16:38"
        },

    // Metadata about the indicator data
    "threat.indicator": {
        "first_seen": "2020-10-01",
        "last_seen": "2020-11-01",
        "sightings": "10",
        "type": [
            "file"
        ],
        "description": "Implant used during an Angler EK campaign.",
        "dataset": "threatintel.malwarebazaar",
        "module": "threatintel",
        "provider": "Abuse.ch",
        "reference": "https://bazaar.abuse.ch/sample/f3ec9a2f2766c6bcf8c2894a9927c227649249ac146aabfe8d26b259be7d7055",
        "confidence": "High",

>         "file": {

Reviewer 2: Apologies, I know I keep circling back to this topic, but I also keep seeing instances of it:

For the indicator use case, would this actually be the top-level `file.*` fields, or nested under `threat.indicator.*` like the example here?

@rylnd @peasead

Reviewer: @ebeahan I agree, I think several of these examples should be moved up a level, e.g.:

* `threat.indicator.dataset` -> `event.dataset`
* `threat.indicator.reference` -> `event.reference`
* `threat.indicator.file.*` -> `file.*`

Author: Thanks, @rylnd and @ebeahan 

After reviewing the [threat enrichment RFC](https://github.com/rylnd/ecs/blob/threat-enrichment-stage-1/rfcs/text/0021-threat-enrichment.md), I understand better. 

I'll make those changes.

Author: @rylnd and @ebeahan 

Shouldn't we move `threat.indicator.domain` to the existing ECS URL fieldset (`url.domain`) as well?

Author: Going with `.domain` being under the existing URL fieldset (`url.domain.`).

> The additions described above will be used to enable cyber threat intelligence capabilities in Elastic Security solution. A new rule type Indicator match will be introduced in 7.10 and the proposed ECS updates will enable a new category of detection alerts that match incoming log and event data against threat intelligence sources. Additionally in the future we will also develop enrichment flows that add context from threat intelligence to alerts and events to assist analysts in their investigative workflows.

Reviewer 2: ```suggestion
The additions described above will be used to enable cyber threat intelligence capabilities in Elastic Security solution. A new rule type Indicator match was introduced in 7.10 and the proposed ECS updates will enable a new category of detection alerts that match incoming log and event data against threat intelligence sources. Additionally in the future we will also develop enrichment flows that add context from threat intelligence to alerts and events to assist analysts in their investigative workflows.
```

 * ECS project (e.g. docs, tooling)
The goal here is to research and understand the impact of these changes on users in the community and development teams across Elastic. 2-5 sentences each.
-->

>  * Ingestion mechanism: Primary ingestion mechanisms will be Filebeat modules and Ingest Packages. There will be no impact on ingestion mechanisms. [Filebeat module](https://www.elastic.co/guide/en/beats/filebeat/7.12/exported-fields-threatintel.html) is scheduled to be released in `7.12`.

Reviewer 2: ```suggestion
 * Ingestion mechanism: Primary ingestion mechanisms will be Filebeat modules and Ingest Packages. There will be no impact on ingestion mechanisms. [Filebeat module](https://www.elastic.co/guide/en/beats/filebeat/7.12/exported-fields-threatintel.html) was released in `7.12`.
```

These sources typically provide intelligence that can be downloaded through REST API or in some cases downloadable CSV's or text files. These intelligence sources will update their data repositories at varying intervals.

> - Abuse.ch Malware - This dataset from Abuse.ch provides a list of malware hashes.

Reviewer 2: The additional details and examples of the intelligence sources are great! 🙌
